### ISSUE TITLE
unit test to test examples

### ISSUE THREAD

**Issue by DanHickstein** [PyAbel/PyAbel#84]
As we are making changes to PyAbel, it seems that we often break the examples. Is it possible to make a unit test that runs the examples? Or, at least the examples that can run quickly (i.e., not basex).

**Comment by rth** [comment]
The fact that we break examples when the interface changes is indeed a problem. Still I do not think that integrating this it in the main test suite, designed to test that PyAbel code works fine is optimal. I think adding a separate test suite would be better (on Travis we can run both one after the other), and I will look into it.

**Comment by DanHickstein** [comment]
That makes sense. A separate test suite would be fine with me. I think that it would help us spot issues with the code, and make sure that the users have reliable examples to start from.

**Comment by MikhailRyazanov** [comment]
If we want to test that the examples just do not fail, rather than that they produce reasonable results, then it might be reasonable to add a test that builds the documentation — it includes all the examples, and checking that the documentation does not break is useful by itself. It does not seem to have any large or unusual dependencies, builds in a reasonable amount of time, and now (after #241) does so without errors and warnings.

**Comment by rth** [comment]
Usually one would just build the documentation in a seperate CI job.

It's not great the examples to the test suite, because the test suite is intended to be fast (TDD development and all that) while examples can take longer to run. Also the way pytest collects test failures etc. is just not great for running examples.

**Comment by MikhailRyazanov** [comment]
I did not mean to include this in the main tests. As you said before, "adding a separate test suite would be better".

Regarding the speed, currently Travis and especially AppVeyor take much more time setting the environment than doing the actual tests. Sometimes AppVeyor takes >10 minutes per test, and it has 6 of them, so this does not look like particularly "fast"...

> Also the way pytest collects test failures etc. is just not great for running examples.

Well, Travis and AppVeyor are not bound to pytest. Not even to Python. So, I suppose, using `make html` would work. Optionally, with the `'warning-is-error'` option for Sphinx to catch the warnings also.

**Comment by MikhailRyazanov** [comment]
By the way, how/when readthedocs builds it? I do not see it being updated after merging #241. Does it need any additional actions? (From you?) Or it has already tried and encountered some fatal error (which I cannot see anywhere)?

**Comment by DanHickstein** [comment]
Circling back to this, I think that we should try to build the documentation on Travis (or appveyor). I'm working on a PR for matplotlib, and they build the documentation and the check fails if there are warnings. It seems like a good idea. I'm not sure exactly how to do it, but I'll look into it further.

**Comment by rth** [comment]
> Circling back to this, I think that we should try to build the documentation on Travis (or appveyor). I'm working on a PR for matplotlib, and they build the documentation and the check fails if there are warnings.

Usually projects do it with CircleCI because it has the ability to store the resulting HTML documentation as an artifact it and serve it for each commit/PR https://circleci.com/blog/deploying-documentation-to-github-pages-with-continuous-integration/ That is what matplotlib is doing for instance, I think.

**Comment by DanHickstein** [comment]
Oh, interesting! Yeah, it would be cool if we could build and display the docs for each PR. This takes care of three things:

1) Tests that the examples build.
2) Test that the docs build
3) Allows us to see the docs and make sure that they look okay. Otherwise this is an annoying process when reviewing a PR that makes changes to the docs.

**Comment by DanHickstein** [comment]
It looks like readthedocs also has a github integration and can serve the documentation for each PR. This looks easier to set up. Does anyone have any preference regarding how we build and display the documentation for each PR?

**Comment by MikhailRyazanov** [comment]
By the way, the “latest” version at readthedocs once again has not updated to the current version (after merging #267, it still corresponds to the previously merged #269).

About the integration:
1. Seeing the documentation built for each PR before it is merged will be very useful! But where it will be available? Not in the default pyabel.readthedocs.io/en/latest, I guess.
2. Is it possible to also keep documentation for several previous releases in addition to just the “latest” (which usually means the current working version, not yet released)? I saw that some other projects have several versions available, so it is possible in principle. Does this require a paid account or something like that?

**Comment by DanHickstein** [comment]
0. I'm not sure why the readthedocs failed. I built the latest version again and it succeeded.

1. I'm not sure exactly how the docs will be available. I don't think that they overwrite the main "latest" build.

2. We can have the docs build for various versions. It was just waiting for me to enable the builds of the previous versions. For example, you can check out the docs for v0.8.3 here:
https://pyabel.readthedocs.io/en/v0.8.3/ and you can use the little green-text selector at the bottom of the window to jump between different versions. Pretty cool! This page shows all of the versions (and, if you're logged in, let's you activate new versions):
https://readthedocs.org/projects/pyabel/versions/

**Comment by MikhailRyazanov** [comment]
Great!

It would be interesting to see how building the documentation for each PR works. And since the documentation now includes almost all examples, this should “test” them also. Which reminds me another question: will we see the build logs?

**Comment by DanHickstein** [comment]
> will we see the build logs
Yeah, I would hope so. I think that anyone can see all the build logs for all of the RTD builds:
https://readthedocs.org/projects/pyabel/builds/

I would hope that it would work in a similar way when we get it set up to autobuild PRs.

I went ahead and sent an email to readthedocs to try to get us set up with the PR-autobuild thing. If we don't like it, we can switch back.

BTW, if you want to play around with all of the settings on our readthedocs account, I think that I can list you as a maintainer. Just make and account on RTD and let me know what your username is. Then I think that I can add you.

**Comment by MikhailRyazanov** [comment]
Now I recall that you already gave me the link to the readthedocs logs... I should save it somewhere. :–)

The failed logs look very strange, like [this one](https://readthedocs.org/projects/pyabel/builds/10344358/):
```
...
ImportError: cannot import name PackageFinder
```
(whereas previous and following builds had no problems with importing that `PackageFinder`).

> if you want to play around with all of the settings on our readthedocs account

Probably not yet. If something strange happens again, I'll look at the logs and poke you if needed. ;–)

Hope to see the autobuilding working soon!

**Comment by DanHickstein** [comment]
Yeah, those errors were strange. I'm not sure why they happened. Seems like a bug with readthedocs.

I think that the autobuild is set up, we should hopefully see it in action for the next PR. (Feel free to open a test PR if you like.)

**Comment by MikhailRyazanov** [comment]
I don't see anything for #270 yet... Will it take more time, or something does not work?

**Comment by DanHickstein** [comment]
Something isn't working right. I e-mailed readthedocs to see what else I should try. I think that we have the webhooks and OAuth stuff enabled, but probably some of the settings aren't quite right.

**Comment by DanHickstein** [comment]
I re-configured the readthedocs webhook to send "Everything" (all activity on the repo), and now it appears to be working! 

I guess I should confirm that breaking an example causes the RTD test to fail...
